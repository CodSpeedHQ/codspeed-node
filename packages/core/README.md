<div align="center">
<h1><code>@codspeed/core</code></h1>

The core Node library used to integrate with Codspeed runners

[![CI](https://github.com/CodSpeedHQ/codspeed-node/actions/workflows/ci.yml/badge.svg)](https://github.com/CodSpeedHQ/codspeed-node/actions/workflows/ci.yml)
[![npm (scoped)](https://img.shields.io/npm/v/@codspeed/core)](https://www.npmjs.com/package/@codspeed/core)
[![Discord](https://img.shields.io/badge/chat%20on-discord-7289da.svg)](https://discord.com/invite/MxpaCfKSqF)

</div>

For now, this crate should not be used directly. Instead, use one of the integration crates:

- `@codspeed/tinybench`
- `@codspeed/benchmark.js`
