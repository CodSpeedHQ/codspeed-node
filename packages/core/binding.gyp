{
  "targets": [
    {
      "target_name": "native_core",
      "cflags!": [
        "-fno-exceptions"
      ],
      "cflags_cc!": [
        "-fno-exceptions"
      ],
      "cflags": [
        "-Wno-maybe-uninitialized",
        "-Wno-unused-variable",
        "-Wno-unused-parameter",
        "-Wno-unused-but-set-variable",
        "-Wno-type-limits"
      ],
      "cflags_cc": [
        "-Wno-maybe-uninitialized",
        "-Wno-unused-variable",
        "-Wno-unused-parameter",
        "-Wno-unused-but-set-variable",
        "-Wno-type-limits"
      ],
      "sources": [
        "src/native_core/linux_perf/linux_perf.cc",
        "src/native_core/linux_perf/linux_perf_listener.cc",
        "src/native_core/instruments/hooks_wrapper.cc",
        "src/native_core/instruments/hooks/dist/core.c",
        "src/native_core/native_core.cc"
      ],
      "include_dirs": [
        "<!@(node -p \"require('node-addon-api').include\")",
        "src/native_core/instruments/hooks/includes"
      ],
      "defines": [
        "NAPI_DISABLE_CPP_EXCEPTIONS"
      ]
    },
  ]
}
