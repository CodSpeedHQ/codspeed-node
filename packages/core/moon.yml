tasks:
  clean:
    args:
      - build
      - generated/openapi
  build:
    deps:
      - "build-native-addon"

  build-native-addon:
    command: prebuildify --napi --strip
    inputs:
      - "src/native_core/**/*.cc"
      - "src/native_core/**/*.h"
      - "binding.gyp"
    outputs:
      - "prebuilds"

  build-tracer-client:
    inputs:
      - "./tracer.spec.json"
    outputs:
      - "src/generated/openapi"
    command: openapi --client axios --input ./tracer.spec.json --name MongoTracer --output ./src/generated/openapi

  typecheck:
    deps:
      - build-tracer-client

  lint:
    deps:
      - build-tracer-client
