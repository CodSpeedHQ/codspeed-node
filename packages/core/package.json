{
  "name": "@codspeed/core",
  "version": "2.1.0",
  "description": "The core Node library used to integrate with Codspeed runners",
  "keywords": [
    "codspeed",
    "benchmark",
    "performance"
  ],
  "files": [
    "dist",
    "prebuilds"
  ],
  "main": "dist/index.cjs.js",
  "module": "dist/index.es5.js",
  "types": "dist/index.d.ts",
  "gypfile": true,
  "author": "Arthur Pastel <arthur@codspeed.io>",
  "repository": "https://github.com/CodSpeedHQ/codspeed-node",
  "homepage": "https://codspeed.io",
  "license": "Apache-2.0",
  "devDependencies": {
    "node-addon-api": "^5.1.0",
    "node-gyp": "^9.3.1",
    "openapi-typescript-codegen": "^0.23.0",
    "prebuildify": "^5.0.1"
  },
  "dependencies": {
    "axios": "^1.4.0",
    "form-data": "^4.0.0",
    "node-gyp-build": "^4.6.0"
  }
}
