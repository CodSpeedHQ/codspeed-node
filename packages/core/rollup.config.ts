import { defineConfig } from "rollup";
import { declarationsPlugin, jsPlugins } from "../../rollup.options";
import pkg from "./package.json" with { type: "json" };

const entrypoint = "src/index.ts";

export default defineConfig([
  {
    input: entrypoint,
    output: [
      {
        file: pkg.types,
        format: "es",
        sourcemap: true,
      },
    ],
    plugins: declarationsPlugin({ compilerOptions: { composite: false } }),
  },
  {
    input: entrypoint,
    output: [
      {
        file: pkg.main,
        format: "cjs",
        sourcemap: true,
      },
      { file: pkg.module, format: "es", sourcemap: true },
    ],
    plugins: jsPlugins(pkg.version),
  },
]);
