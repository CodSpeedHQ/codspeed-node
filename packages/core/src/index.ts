import { tryIntrospect } from "./introspection";
tryIntrospect();

import native_core from "./native_core";
import { initOptimization } from "./optimization";

declare const __VERSION__: string;

const linuxPerf = new native_core.LinuxPerf();

export const isBound = native_core.isBound;

export const setupCore = () => {
  initOptimization();
  native_core.Measurement.stopInstrumentation(
    `Metadata: codspeed-node ${__VERSION__}`
  );
  linuxPerf.start();
};

export const teardownCore = () => {
  linuxPerf.stop();
};

export { optimizeFunction, optimizeFunctionSync } from "./optimization";
export const Measurement = native_core.Measurement;
