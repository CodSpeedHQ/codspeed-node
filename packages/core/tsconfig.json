{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "declaration": true,
    "outDir": "dist/types",
    "rootDir": "src"
  },
  "include": [
    "src/**/*.ts"
  ]
}
