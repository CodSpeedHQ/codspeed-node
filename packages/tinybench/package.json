{
  "name": "@codspeed/tinybench",
  "version": "0.0.0",
  "description": "tinybench compatibility layer for CodSpeed",
  "keywords": [
    "codspeed",
    "benchmark",
    "tinybench",
    "performance"
  ],
  "main": "dist/index.cjs.js",
  "module": "dist/index.es5.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "author": "Arthur Pastel <arthur@codspeed.io>",
  "repository": "https://github.com/CodSpeedHQ/codspeed-node/tree/main/packages/tinybench",
  "homepage": "https://codspeed.io",
  "license": "MIT",
  "devDependencies": {
    "@types/find-up": "^4.0.0",
    "@types/stack-trace": "^0.0.30",
    "jest-mock-extended": "^3.0.1"
  },
  "dependencies": {
    "@codspeed/core": "workspace:*",
    "find-up": "^6.3.0",
    "stack-trace": "1.0.0-pre1",
    "tinybench": "^2.3.1"
  }
}
