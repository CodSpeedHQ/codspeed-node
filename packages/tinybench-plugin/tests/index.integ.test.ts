import { mockDeep, mockReset } from "jest-mock-extended";
const mockCore = mockDeep<typeof core>();

import * as core from "@codspeed/core";
import { Bench } from "tinybench";
import { withCodSpeed } from "..";
import { registerBenchmarks } from "./registerBenchmarks";
import { registerOtherBenchmarks } from "./registerOtherBenchmarks";

jest.mock("@codspeed/core", () => mockCore);

beforeEach(() => {
  mockReset(mockCore);
  jest.clearAllMocks();
});

describe("Benchmark.Suite", () => {
  it("simple suite", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(false);
    const bench = withCodSpeed(new Bench({ time: 100 }));
    const onComplete = jest.fn();
    bench.add("RegExp", function () {
      /o/.test("Hello World!");
    });
    bench.getTask("RegExp")?.addEventListener("complete", onComplete);
    await bench.run();

    expect(onComplete).toHaveBeenCalled();
    expect(mockCore.Measurement.startInstrumentation).not.toHaveBeenCalled();
    expect(mockCore.Measurement.stopInstrumentation).not.toHaveBeenCalled();
  });
  it("check core methods are called", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(true);
    await withCodSpeed(new Bench())
      .add("RegExp", function () {
        /o/.test("Hello World!");
      })
      .run();
    expect(mockCore.Measurement.startInstrumentation).toHaveBeenCalled();
    expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/index.integ.test.ts::RegExp"
    );
  });
  it("check suite name is in the uri", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(true);
    await withCodSpeed(new Bench())
      .add("RegExp", function () {
        /o/.test("Hello World!");
      })
      .add("RegExp2", () => {
        /o/.test("Hello World!");
      })
      .run();
    expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/index.integ.test.ts::RegExp"
    );
    expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/index.integ.test.ts::RegExp2"
    );
  });
  it("check error handling", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(true);
    const bench = withCodSpeed(new Bench());
    bench.add("throwing", async () => {
      throw new Error("test");
    });
    await expect(bench.run()).rejects.toThrowError("test");
  });
  it.each([true, false])(
    "check console output(instrumented=%p) ",
    async (instrumented) => {
      const logSpy = jest.spyOn(console, "log");
      const warnSpy = jest.spyOn(console, "warn");
      mockCore.Measurement.isInstrumented.mockReturnValue(instrumented);
      await withCodSpeed(new Bench({ time: 100 }))
        .add("RegExp", function () {
          /o/.test("Hello World!");
        })
        .add("RegExp2", () => {
          /o/.test("Hello World!");
        })
        .run();
      // Check that the first log contains "[CodSpeed] running with @codspeed/tinybench v"
      if (instrumented) {
        expect(logSpy).toHaveBeenCalledWith(
          expect.stringContaining(
            "[CodSpeed] running with @codspeed/tinybench v"
          )
        );
        expect({
          log: logSpy.mock.calls.slice(1),
          warn: warnSpy.mock.calls,
        }).toMatchSnapshot();
      } else {
        expect({
          log: logSpy.mock.calls,
          warn: warnSpy.mock.calls,
        }).toMatchSnapshot();
      }
    }
  );
  it("check nested file path is in the uri when bench is registered in another file", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(true);
    const bench = withCodSpeed(new Bench());
    registerBenchmarks(bench);
    await bench.run();
    expect(mockCore.Measurement.startInstrumentation).toHaveBeenCalled();
    expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
      "packages/tinybench-plugin/tests/registerBenchmarks.ts::RegExp"
    );
  });
  // TODO: this is not supported at the moment as tinybench does not support tasks with same name
  // remove `.failing` when tinybench supports it
  it.failing(
    "check that benchmarks with same name have different URIs when registered in different files",
    async () => {
      mockCore.Measurement.isInstrumented.mockReturnValue(true);
      const bench = withCodSpeed(new Bench());
      registerBenchmarks(bench);
      registerOtherBenchmarks(bench);
      await bench.run();
      expect(mockCore.Measurement.startInstrumentation).toHaveBeenCalled();
      expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
        "packages/tinybench-plugin/tests/registerBenchmarks.ts::RegExp"
      );
      expect(mockCore.Measurement.stopInstrumentation).toHaveBeenCalledWith(
        "packages/tinybench-plugin/tests/registerOtherBenchmarks.ts::RegExp"
      );
    }
  );

  it("should run before and after hooks", async () => {
    mockCore.Measurement.isInstrumented.mockReturnValue(true);
    const beforeAll = jest.fn();
    const beforeEach = jest.fn();
    const afterEach = jest.fn();
    const afterAll = jest.fn();

    await withCodSpeed(new Bench())
      .add(
        "RegExp",
        function () {
          /o/.test("Hello World!");
        },
        { afterAll, afterEach, beforeAll, beforeEach }
      )
      .add(
        "RegExp2",
        () => {
          /o/.test("Hello World!");
        },
        { afterAll, afterEach, beforeAll, beforeEach }
      )
      .run();

    expect(beforeAll).toHaveBeenCalledTimes(2);
    expect(beforeEach).toHaveBeenCalledTimes(2);
    expect(afterEach).toHaveBeenCalledTimes(2);
    expect(afterAll).toHaveBeenCalledTimes(2);
  });
});
