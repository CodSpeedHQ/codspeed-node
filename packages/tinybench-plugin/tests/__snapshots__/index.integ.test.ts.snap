// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Benchmark.Suite > check console output(instrumented=%p)  false 1`] = `
{
  "log": [],
  "warn": [
    [
      "[CodSpeed] 2 benches detected but no instrumentation found, falling back to tinybench",
    ],
  ],
}
`;

exports[`Benchmark.Suite > check console output(instrumented=%p)  true 1`] = `
{
  "log": [
    [
      "    ✔ Measured packages/tinybench-plugin/tests/index.integ.test.ts::RegExp",
    ],
    [
      "    ✔ Measured packages/tinybench-plugin/tests/index.integ.test.ts::RegExp2",
    ],
    [
      "[CodSpeed] Done running 2 benches.",
    ],
  ],
  "warn": [],
}
`;
